modules/printable_pdf/src/Plugin/PrintableFormat/PdfFormat.php

```diff
@@ -0,0 +1,205 @@
+    parent::__construct($configuration,$plugin_id, $plugin_definition, $config_factory, $printable_css_include,$link_extractor);
+    $config = $this->getConfiguration();
+    $this->pdfGeneratorManager = $pdf_generator_manager;
+    $pdf_library = (string)$this->configFactory->get('printable.settings')->get('pdf_tool');
+    if($pdf_library == $config['pdf_generator'])
+      $this->pdfGenerator = $this->pdfGeneratorManager->createInstance($config['pdf_generator']);
+    else if($pdf_library == 'mPDF')
+      $this->pdfGenerator = $this->pdfGeneratorManager->createInstance('mpdf');
+    else
+      $this->pdfGenerator = $this->pdfGeneratorManager->createInstance('tcpdf');
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+    return new static(
+      $configuration, $plugin_id, $plugin_definition,
+      $container->get('config.factory'),
+      $container->get('plugin.manager.pdf_generator'),
+      $container->get('printable.css_include'),
+      $container->get('printable.link_extractor')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function defaultConfiguration() {
+    return array(
+      'pdf_generator' => 'wkhtmltopdf'
+    );
+  }
+  public function calculateDependencies(){}
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildConfigurationForm(array $form, FormStateInterface $form_state) {
+    $config = $this->getConfiguration();
+    $options = array();
+    foreach ($this->pdfGeneratorManager->getDefinitions() as $definition) {
+      $options[$definition['id']] = $definition['title'];
+    }
+    $form['pdf_generator'] = array(
+      '#type' => 'radios',
+      '#title' => 'PDF Generator',
+      '#default_value' => $config['pdf_generator'],
+      '#options' => $options,
+    );
+
+    return $form;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function validateConfigurationForm(array &$form, FormStateInterface $form_state){}
+
+
+  /**
+   * {@inheritdoc}
+   */
+  public function submitConfigurationForm(array &$form, FormStateInterface $form_state) {
+    $this->setConfiguration(array(
+      'pdf_generator' => $form_state['values']['pdf_generator'],
+    ));
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildContent() {
+   $content = parent::buildContent();
+   $rendered_page = parent::extractLinks(render($content));
+   $this->pdfGenerator->addPage($rendered_page);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function mbuildContent($save_pdf, $filename) {
+    $content = parent::buildContent();
+    $rendered_page = parent::extractLinks(render($content));
+    if($save_pdf) {
+      if(empty($filename)) {
+        $filename=str_replace("/","_",\Drupal::service('path.current')->getPath());
+        $filename=substr($filename, 1);
```

Comment:
What does this do? strip a trailing character? Then use trim() that explains itself.